testdata/p4_16_samples/pna-direction-main-parser-err.p4

```diff
@@ -0,0 +1,185 @@
+
+struct empty_metadata_t {
+}
+
+//////////////////////////////////////////////////////////////////////
+// Struct types for holding user-defined collections of headers and
+// metadata in the P4 developer's program.
+//
+// Note: The names of these struct types are completely up to the P4
+// developer, as are their member fields, with the only restriction
+// being that the structs intended to contain headers should only
+// contain members whose types are header, header stack, or
+// header_union.
+//////////////////////////////////////////////////////////////////////
+
+struct main_metadata_t {
+    // empty for this skeleton
+    bit<32> tmpDir;
+}
+
+// User-defined struct containing all of those headers parsed in the
+// main parser.
+struct headers_t {
+    ethernet_t ethernet;
+    ipv4_t ipv4;
+}
+
+control PreControlImpl(
+    in    headers_t  hdr,
+    inout main_metadata_t meta,
+    in    pna_pre_input_metadata_t  istd,
+    inout pna_pre_output_metadata_t ostd)
+{
+
+    apply {
+        if (PNA_Direction_t.NET_TO_HOST == istd.direction) {
+            meta.tmpDir = hdr.ipv4.srcAddr;
+        } else {
+            meta.tmpDir = hdr.ipv4.dstAddr;
+        }
+
+    }
+
+}
+
+parser MainParserImpl(
+    packet_in pkt,
+    out   headers_t       hdr,
+    inout main_metadata_t main_meta,
+    in    pna_main_parser_input_metadata_t istd)
+{
+    bit<32> tmpDir;
+
+    state start {
+        pkt.extract(hdr.ethernet);
+        transition select(hdr.ethernet.etherType) {
+            0x0800: parse_ipv4;
+            default: accept;
+        }
+    }
+    state parse_ipv4 {
+        pkt.extract(hdr.ipv4);
+        if (PNA_Direction_t.NET_TO_HOST == istd.direction) {
+            tmpDir = hdr.ipv4.srcAddr;
+        } else {
+            tmpDir = hdr.ipv4.dstAddr;
+        }
+
+        transition accept;
+    }
+}
+
+//Register<bit<32>, bit<16>>(1) direction_port_mask;
+control MainControlImpl(
+    inout headers_t       hdr,           // from main parser
+    inout main_metadata_t user_meta,     // from main parser, to "next block"
+    in    pna_main_input_metadata_t  istd,
+    inout pna_main_output_metadata_t ostd)
+{
+    bit<32> tmpDir;
+// This is to be inserted by the compiler for direction metadata
+   //Register<bit<32>, bit<16>>(1) direction_port_mask;
+    action next_hop(PortId_t vport) {
+        send_to_port(vport);
+    }
+    action default_route_drop() {
+        drop_packet();
+    }
+    table ipv4_da_lpm {
+        key = {
+            tmpDir :lpm @name ("ipv4_addr");
+        }
+        actions = {
+            next_hop;
+            default_route_drop;
+        }
+        const default_action = default_route_drop;
+    }
+    apply {
+
+        if (PNA_Direction_t.NET_TO_HOST == istd.direction) {
+            tmpDir = hdr.ipv4.srcAddr;
+        } else {
+            tmpDir = hdr.ipv4.dstAddr;
+        }
+/*
+//        istd.direction should be replaced with "direction_port_mask.read(0) & (32w0x1 << istd.input_port"
```

Comment:
You can remove this commented code now. This was just for explaining the implementation detail.